File: spring.py
# uncomment the next line if running in a notebook
# %matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
from math import sqrt, cos, sin


def do_analytic(k_0, m_0, x_0, v_0):

    w = sqrt(k_0 / m_0)
    x = x_0
    v = v_0

    x_list = []
    v_list = []

    for t in t_array:

        x_list.append(x)
        v_list.append(v)

        x = (1 / w) * sin(w * t)
        v = cos(w * t)

    # convert trajectory lists into arrays, so they can be sliced (useful for Assignment 2)
    x_array = np.array(x_list)
    v_array = np.array(v_list)
    return (x_array, v_array)


def do_integration(m_0, k_0, x_0, v_0, func):

    m = m_0
    k = k_0
    x = x_0
    v = v_0
    # initialise empty lists to record trajectories
    x_list = []
    v_list = []

    # Euler integration
    for t in t_array:

        # append last state to trajectories
        x_list.append(x)
        v_list.append(v)

        # calculate new position and velocity for next iteration
        if t == 0:
            a, x, v = do_euler(m, k, x, v)
        else:
            a, x, v = func(m, k, x, v, x_list[-2])

    # convert trajectory lists into arrays, so they can be sliced (useful for Assignment 2)
    x_array = np.array(x_list)
    v_array = np.array(v_list)
    return (x_array, v_array)


def do_euler(m, k, x, v, *argv):
    # calculate new position and velocity for next iteration
    a = -k * x / m
    x = x + dt * v
    v = v + dt * a
    return (a, x, v)


def do_verlet(m, k, x, v, *argv):
    px = x
    # calculate new position and velocity for next iteration
    a = -k * x / m
    x = 2 * x - argv[0] + (dt ** 2) * a
    v = (x - px) / dt
    return (a, x, v)


# mass, spring constant, initial position and velocity
m = 1
k = 1
x = 0
v = 1

# simulation time, timestep and time
t_max = 100
dt = 0.1
t_array = np.arange(0, t_max, dt)

eulers = do_integration(m, k, x, v, do_euler)
verlets = do_integration(m, k, x, v, do_verlet)
analytics = do_analytic(m, k, x, v)

# plot the position-time graph
plt.figure(1)
plt.clf()
plt.xlabel("time (s)")
plt.grid()
plt.plot(t_array, eulers[0], label="x (m)")
plt.plot(t_array, eulers[1], label="v (m/s)")
plt.plot(t_array, verlets[0], label="x (m) verlet")
plt.plot(t_array, verlets[1], label="v (m/s) verlet")
plt.plot(t_array, analytics[0], label="x (m) anayl")
plt.plot(t_array, analytics[1], label="v (m/s) anayl")
plt.legend()
plt.show()

File: spring.cpp
#include <fstream>
#include <iostream>
#include <vector>

#define SIMULATION_TIME 100

void integrate(struct Integrator* integrator, double dt);
void do_euler(struct Integrator* integrator, double dt);
void do_verlet(struct Integrator* integrator, double dt, double ppx);

using namespace std;

struct Integrator {
  double m;
  double k;
  double x_0;
  double v_0;
  vector<double> t_list;
  vector<double> x_list;
  vector<double> v_list;

  // temp vars to be stored in the struct for calcs
  double a;
  double x;
  double v;

  // config
  bool isVerlet;
};

void integrate(struct Integrator* integrator, double dt) {
  // Euler integration
  double t;
  integrator->x = integrator->x_0;
  integrator->v = integrator->v_0;
  for (t = 0; t <= SIMULATION_TIME; t = t + dt) {
    // append current state to trajectories
    double ppx;
    if (!integrator->x_list.empty()) {
      ppx = integrator->x_list.back();
    }

    integrator->t_list.push_back(t);
    integrator->x_list.push_back(integrator->x);
    integrator->v_list.push_back(integrator->v);

    if (integrator->x_list.empty() && integrator->isVerlet) {
      do_euler(integrator, dt);
    } else if (integrator->isVerlet) {
      do_verlet(integrator, dt, ppx);
    } else {
      do_euler(integrator, dt);
    }
  }
}

void do_euler(struct Integrator* integrator, double dt) {
  // calculate new position and velocity
  struct Integrator* i = integrator;
  i->a = -i->k * i->x / i->m;
  i->x = i->x + dt * i->v;
  i->v = i->v + dt * i->a;
};

void do_verlet(struct Integrator* integrator, double dt, double ppx) {
  // calculate new position and velocity
  struct Integrator* i = integrator;
  i->a = -i->k * i->x / i->m;
  double px = i->x;
  i->x = 2 * i->x - ppx + (dt * dt) * i->a;
  i->v = (i->x - px) / dt;
};

void save_trajectory(struct Integrator* integrator) {
  // Write the trajectories to file
  ofstream fout;
  if (integrator->isVerlet) {
    fout.open("trajectories-verlet.txt");

  } else {
    fout.open("trajectories-euler.txt");
  }
  if (fout) {  // file opened successfully
    for (int i = 0; i < integrator->t_list.size(); i = i + 1) {
      fout << integrator->t_list[i] << ' ' << integrator->x_list[i] << ' '
           << integrator->v_list[i] << endl;
    }
  } else {  // file did not open successfully
    cout << "Could not open trajectory file for writing" << endl;
  }
}

int main() {
  // declare variables
  struct Integrator euler_integrator;
  struct Integrator verlet_integrator;

  // mass, spring constant, initial position and velocity
  euler_integrator.m = 1;
  euler_integrator.k = 1;
  euler_integrator.x_0 = 0;
  euler_integrator.v_0 = 1;
  euler_integrator.isVerlet = false;

  verlet_integrator.m = 1;
  verlet_integrator.k = 1;
  verlet_integrator.x_0 = 0;
  verlet_integrator.v_0 = 1;
  verlet_integrator.isVerlet = true;

  integrate(&euler_integrator, 0.1);
  integrate(&verlet_integrator, 0.1);

  save_trajectory(&euler_integrator);
  save_trajectory(&verlet_integrator);
}
